Guitar Chord Diagrams

charles-l opened the pull request:
What does your Instant Answer do?
It returns a requested guitar chord diagram.

What problem does your Instant Answer solve (Why is it better than organic links)?
Easier access.

What is the data source for your Instant Answer? (Provide a link if possible)
The GD::Tab::Guitar cpan module. (http://search.cpan.org/~taniguchi/GD-Tab-Guitar-0.03/lib/GD/Tab/Guitar.pm)

Why did you choose this data source?
Because it generates the diagrams.

Are there any other alternative (better) data sources?
Not that I know of.

What are some example queries that trigger this Instant Answer?
'G guitar chord' 'G#m6 guitar chord'

Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)
Musicians.

Is this Instant Answer connected to a DuckDuckHack Instant Answer idea?
(https://duck.co/ideas/idea/537/piano-guitar-chords), but it looks like it was abandoned?

Which existing Instant Answers will this one supercede/overlap with?
None.

Are you having any problems? Do you need our help with anything?
Nope.

What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)
![screen shot 2014-12-19 at 7 07 51 pm](https://cloud.githubusercontent.com/assets/1291012/5512936/5f161618-87b2-11e4-8327-a4b6f0ca8e2d.png)


[x] Added metadata and attribution information (https://duck.co/duckduckhack/metadata)
[x] Wrote test file and added to t/ directory (https://duck.co/duckduckhack/goodie_tests)
[x] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[x] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[x] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 8
        [] IE 9
        [] IE 10

    - Windows XP
        [] IE 7
        [] IE 8

    - Mac OSX
        [x] Google Chrome
        [x] Firefox
        [] Opera
        [x] Safari

    - iOS (iPhone)
        [x] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [] Firefox
        [] Native Browser
        [] Google Chrome
        [] Opera

mintsoft commented:
@charles-l I don't think there's a problem with the chord name being in the image; it's part of the chord-diagram imo

mintsoft commented:
@chrismorast any ideas for the design?
@moollaza anything else to add?

chrismorast commented:
Nice @charles-l , this is a great idea!

Is this an image being generated? Do we have any freedom with it (ie. size, color, etc.)?

charles-l commented:
@chrismorast thanks! The images are generated manually beforehand, but you do have some options in the script for what it looks like (size and color). There's no way of modifying the image on the fly, though.

chrismorast commented:
Is it possible to try going about 150-200% in size (of what it is now) and use the color # 333. As far as the font goes, can we use Proxima Nova weight of 600? Not sure is css classes would work here but if so I can get class names for you.

jagtalon commented:
@charles-l Nice!

- Will adding `before_install = sudo apt-get install -y libgd-dev` in `dist.ini` work? https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/dist.ini#L141

jagtalon commented:
Looking forward to the bigger image! Just ping us if you need any help. :)

charles-l commented:
@chrismorast @jagtalon OK, I updated the style a little. I had to convert the .woff version of the font to a .ttf and put it in `share/`. I had to modify the original script to allow for a custom font, so I've copied the script to the `share/` and added a the required tweaks.

![screen shot 2015-01-15 at 1 04 38 pm](https://cloud.githubusercontent.com/assets/1291012/5763275/17b800a0-9cb7-11e4-9663-b8adb42a1952.png)

@jagtalon I've added the install command in `dist.ini`. It is still failing, because the images aren't generated for tests. How would I run a make command before a test?

mintsoft commented on lib/DDG/Goodie/GuitarChords.pm:
```diff
@@ -0,0 +1,78 @@
+
+# guitar script is stored in share directory
+# including this way, because it limits duplicated code
+my $g = share("Guitar.pm");
```
@charles-l I thought the images would be pre-generated and committed into /share and then included with the ImageLoaderRole; that way it wouldn't require GD for live or the tests

charles-l replied:
@mintsoft ah, that's a fine route, too. I just didn't want to gum up the repo with a bunch of binary files. Bear in mind, however that it generates ~500 images. They're not big (the size of the directory is 2.1M after they're generated), but that's still a few hundred lines to scroll through when browsing the repo.

mintsoft replied:
@charles-l I see your concern; personally I don't really have a problem with that. @moollaza @jagtalon any issues with this approach?

jagtalon replied:
@charles-l I like that approach. A few questions, though:

- Would it only read in one image or everything?
- Do the images work for high-res screens like Retina?

charles-l replied:
@jagtalon The makefile will generate all possible guitar chords into the share folder all at once.

I'm not generating high-res images at the moment. How would I handle checking that the client needs a high resolution image? It wouldn't really make sense to put it in the CSS (as the image is embedded as base64). Is there some way of getting that from the perl script?

jagtalon replied:
I see. It is a bit blurry, but please generate the images for us. :)

jagtalon replied:
![img_0563](https://cloud.githubusercontent.com/assets/81969/5843484/d42e16e4-a17a-11e4-9bc5-9d8aa45ffd89.PNG)

jagtalon commented on dist.ini:
```diff
@@ -137,3 +137,4 @@ extra_dep = autodie
 after_install = duckpan DDG
 before_install = sudo apt-get update -qq
 before_install = sudo apt-get install -y libmpfr-dev
+before_install = sudo apt-get install -y libgd-dev
```
Shouldn't GD be added in here too?

jagtalon replied:
Looks like this isn't the right package name?

jagtalon commented:
@charles-l Alright, so since it's blurry on retina screens, it would be great if we could have the generated images as hi-res images instead. It would be hi-res everywhere, but I think that's okay. :) What do you think?

jkanarek commented:
Is it possible to return "related" chords as well (i.e. other G# chords), might be useful!

charles-l commented:
@jagtalon yep, that sounds good. I'll update the script to generate higher res images (is there a specific size that would be better, or should I just double the resolution?)

@jkanarek like alternative versions of the chord? I could add something to the script, but I may have to have a list of chords already. I'll look into that.

jagtalon commented:
@charles-l double should be fine. :)

charles-l commented:
@jagtalon sorry for the delay. I'm getting an error when I try to upgrade to the latest version of duckpan (and as a result can't run the server to test the goodie). Is there some way of forcing the server to start even if it's not the latest version?

moollaza commented:
@charles-l what error are you seeing? I just released a new version of DuckPAN which should have fixed any recent issues

charles-l commented:
@moollaza this is the error in `duckpan upgrade`:
```  
    #   Failed test 'setting DuckPAN env test to me'
    #   at /home/dax/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/Test/Exception.pm line 346.
    #          got: 2
    #     expected: 0
Error open (<) on '/tmp/E83jjRbBra/env.ini': No such file or directory at t/system_duckpan.t line 39.
    # Child (env) exited without calling finalize()

#   Failed test 'env'
#   at /home/dax/perl5/perlbrew/perls/perl-5.16.3/lib/site_perl/5.16.3/Test/Builder.pm line 279.
# Tests were run but no plan was declared and done_testing() was not seen.
# Looks like your test exited with 2 just after 3.
t/system_duckpan.t ...... 
Dubious, test returned 2 (wstat 512, 0x200)
Failed 1/3 subtests 
```
I'm using the virtualbox vm, so I think I have all the correct dependencies. Does `t/system_duckpan.t` rely on any new external packages?

moollaza commented:
> Does t/system_duckpan.t rely on any new external packages?

We have new tests which use Path::Tiny, but it's in the dist, so the recent upgrade should have installed it, if it wasn't already. I previously experienced a very similar issue with these tests (creating temp files) and couldn't figure out the cause. When I tested again after the break, the problem was no longer occurring on any machines so it seemed safe to release. For now I might have to remove the tests and re-release.

charles-l commented:
@moollaza OK, is there something I can do to not run any tests when I install it?

moollaza commented:
@charles-l a new version of DuckPAN has been released which doesn't include the failing tests. You should be able to install that without any problems

charles-l commented:
@moollaza Awesome, the upgrade worked.

@jagtalon OK, generated images (and removed external deps). They're double the resolution now, and scaled down to half size, so I think it'll look good on retina :)

moollaza commented:
>Awesome, the upgrade worked.

@charles-l that's great to hear! Sorry about the troubles.

@jagtalon @charles-l I'm a little out of the loop on this, where exactly are we on this one? Ready for signoffs and merge?

mintsoft commented:
@moollaza Awaiting design/DDG sign-off/ and merge :)

jagtalon commented:
@charles-l Nice!

- It looks like the tests are failing for me when I ran `prove -Ilib t/GuitarChords.t`
- Since `chords.pm` is just a hash, maybe we should just move it inside GuitarChords.pm?

jagtalon commented:
Oh another one:

- Can you update the test with queries that **don't** trigger this IA. Maybe things like `C# programming` might be a start. Here's an example of that: https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/t/Calculator.t#L680

jagtalon commented:
@charles-l could the images with `(` be escaped somehow? It looks like bash doesn't like it so we can't do a `dzil install`: 

```
...
Checking if you have Text::UpsideDown 1.22 ... Yes (1.22)
Checking if you have Number::UN 0.002 ... Yes (0.003)
Checking if you have Net::IDN::Encode 2.003 ... Yes (2.100)
Building and testing DDG-GoodieBundle-OpenSourceDuckDuckGo-1051
/bin/sh: 2: Syntax error: "(" unexpected
make: *** [config] Error 2
-> FAIL Installing . failed. See /home/ubuntu/.cpanm/work/1423218705.2451/build.log for details. Retry with --force to force install it.
```

mintsoft commented:
@jagtalon I don't understand what you mean here :)

charles-l commented:
@mintsoft that's my bad, the images are generated to the share directory, but the file names have '(' in them, so when dzil install finds them, they aren't escaped and it fails.

charles-l commented:
@jagtalon @mintsoft I think everything should be working now :)

charles-l commented:
Huh. The second build failed for some reason...

mintsoft commented on lib/DDG/Goodie/GuitarChords.pm:
```diff
@@ -0,0 +1,79 @@
+        $r .= $d if $d;
+        $r .= $e if $e;
+        $r .= $f if $f;
+        if ($r ~~ @{[keys %chord_lists]}) {
```
This is what's failed the build:
https://travis-ci.org/duckduckgo/zeroclickinfo-goodies/jobs/50566417#L3675

charles-l replied:
@mintsoft err... sorry, I'm still missing it. It's saying something about a TAP plan? Dunno what that is, though.

moollaza replied:
@charles-l you need to scroll up a little to line 3675:

```
[23:52:22] t/GuitarChords.t ................. Smartmatch is experimental at /home/travis/build/duckduckgo/zeroclickinfo-goodies/.build/9vst6yMqvN/blib/lib/DDG/Goodie/GuitarChords.pm line 67.

Compilation failed in require at t/GuitarChords.t line 12.

[23:52:22] t/GuitarChords.t ................. Dubious, test returned 255 (wstat 65280, 0xff00)

No subtests run 
```

The problem is the "smartmatch" operator (`~~`), which I guess is considered experimental. Any chance you can find an alternative to that?

mintsoft commented:
@charles-l I fixed the most recent build prolem yesterday; can you merge-in again?

charles-l commented:
@mintsoft merged, and it works! :D

mintsoft commented:
@charles-l Good news then! @jagtalon @moollaza I think this is ready

moollaza commented:
Sorry about the delay guys, will try and get around to this one soon!

charles-l commented:
@moollaza no problem, and thanks for the heads up :)

jagtalon commented:
@charles-l thanks! Putting it up on a server.

jagtalon commented:
@abeyang up on https://ddh9.duckduckgo.com/?q=G%23m6+guitar+chord&ia=answer! @moollaza I don't know much about guitar chords--maybe you could spot check the IA to see if the info is correct?

jagtalon commented:
@charles-l thanks for fixing the filenames btw!

jagtalon commented on lib/DDG/Goodie/GuitarChords.pm:
```diff
@@ -0,0 +1,79 @@
+    return;
+};
+
+sub check_chord
```
Could you add a brief comment before the function which explains what it does? It's a bit intimidating at first glance.

jagtalon commented:
:boom:

moollaza commented:
@charles-l Thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's & Issues or even submit more Instant Answers!

charles-l commented:
@jagtalon, @moollaza Awesome! I added some extra comments to the code.

@moollaza Yeah, I do plan on writing more instant answers soon ;)

mintsoft commented:
Congrats @charles-l :tada:

charles-l commented:
@mintsoft thank you

And thank you for all the help :)

mintsoft commented:
Happy to help @charles-l !

moollaza commented:
@charles-l congrats! It's live: https://duckduckgo.com/?q=G+guitar+chord&ia=answer

Thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's & Issues or even submit more Instant Answers!
